File: server/services/auth/internal/csrf/csrf.go
// Package csrf implements the double-submit cookie CSRF defence with a
// strict, exact-match origin allowlist. It intentionally contains no
// substring or "contains" host matching — that class of check is bypassable.
package csrf

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type SecretStore interface {
	Save(ctx context.Context, sessionIDHash, secretHash string, expiresAt time.Time) error
	Find(ctx context.Context, sessionIDHash string) (secretHash string, expiresAt time.Time, err error)
}

func hashToken(v string) string {
	sum := sha256.Sum256([]byte(v))
	return base64.RawStdEncoding.EncodeToString(sum[:])
}

// HashForLookup exposes the token hashing used for server-side binding keys.
func HashForLookup(v string) string { return hashToken(v) }

// NewSecret generates 256 bits of URL-safe entropy.
func NewSecret() (string, error) {
	b := make([]byte, 32)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(b), nil
}

// Allowlist holds the exact origins permitted to make state-changing browser
// requests.
type Allowlist struct {
	origins map[string]struct{}
}

func NewAllowlist(publicURL string, extra []string) *Allowlist {
	a := &Allowlist{origins: make(map[string]struct{})}
	for _, o := range append([]string{publicURL}, extra...) {
		o = strings.TrimSuffix(strings.TrimSpace(o), "/")
		if o == "" {
			continue
		}
		if !strings.Contains(o, "://") {
			continue // origins must include scheme; bare hosts are rejected
		}
		if u, err := url.Parse(o); err == nil && u.Host != "" {
			a.origins[u.Scheme+"://"+u.Host] = struct{}{}
		}
	}
	return a
}

func (a *Allowlist) Contains(origin string) bool {
	_, ok := a.origins[strings.TrimSuffix(origin, "/")]
	return ok
}

// OriginAllowed applies the request-side rules:
//
//  1. An explicit Origin header must exactly match the allowlist OR the
//     request's own Host-derived origin (same-origin deployments).
//  2. Browsers always attach Origin to cross-site and same-site POSTs; a
//     mutating request WITH credentials but NO Origin and NO Referer is
//     treated as hostile (fail closed).
//  3. Non-browser clients without cookies are unaffected (they cannot be
//     CSRF'd; they authenticate explicitly).
func OriginAllowed(r *http.Request, allow *Allowlist) bool {
	origin := r.Header.Get("Origin")
	referer := r.Header.Get("Referer")

	scheme := "http"
	if r.TLS != nil || strings.EqualFold(r.Header.Get("X-Forwarded-Proto"), "https") {
		scheme = "https"
	}
	selfOrigin := scheme + "://" + r.Host

	match := func(candidate string) bool {
		return candidate == selfOrigin || allow.Contains(candidate)
	}

	if origin != "" {
		return match(strings.TrimSuffix(origin, "/"))
	}
	if referer != "" {
		if u, err := url.Parse(referer); err == nil && u.Host != "" {
			return match(u.Scheme + "://" + u.Host)
		}
		return false
	}

	// No Origin and no Referer. Modern browsers send Sec-Fetch-Site on every
	// fetch/navigation; if present this IS a browser request → reject.
	if r.Header.Get("Sec-Fetch-Site") != "" {
		return false
	}
	// A cookie-bearing mutation without any origin signal is rejected too.
	if hasSessionishCookie(r) {
		return false
	}
	return true
}

func hasSessionishCookie(r *http.Request) bool {
	_, err := r.Cookie("meshwork_session")
	if err == nil {
		return true
	}
	_, err = r.Cookie("__Host-meshwork_session")
	return err == nil
}

// Verify performs the double-submit comparison plus optional server-side
// session binding. cookieToken and headerToken come from the client;
// sessionIDHash (optional) enables bound-secret verification.
func Verify(cookieToken, headerToken, sessionIDHash string, secrets SecretStore) bool {
	if cookieToken == "" || headerToken == "" || len(cookieToken) > 512 || len(headerToken) > 512 {
		return false
	}
	if subtle.ConstantTimeCompare([]byte(cookieToken), []byte(headerToken)) != 1 {
		return false
	}
	if sessionIDHash == "" || secrets == nil {
		return true // unauthenticated endpoints rely on double-submit alone
	}
	storedHash, expiresAt, err := secrets.Find(context.Background(), hashToken(sessionIDHash))
	if err != nil || expiresAt.Before(time.Now()) {
		// No bound record yet — bind this token now (first use after login).
		_ = secrets.Save(context.Background(), hashToken(sessionIDHash), hashToken(headerToken),
			time.Now().Add(time.Hour))
		return true
	}
	return subtle.ConstantTimeCompare([]byte(storedHash), []byte(hashToken(headerToken))) == 1
}

// Bind stores the secret for the session so subsequent requests verify
// against the server-side copy (defence against cookie injection).
func Bind(secrets SecretStore, sessionIDHash, secret string, ttl time.Duration) error {
	if sessionIDHash == "" || secrets == nil {
		return nil
	}
	return secrets.Save(context.Background(), hashToken(sessionIDHash), hashToken(secret), time.Now().Add(ttl))
}

File: server/auth/csrf.ts
// CSRF defence for monolith-served mutations: strict exact-match origin
// allowlist + double-submit cookie comparison. The DB-bound secret check is
// intentionally NOT duplicated here — the Go auth service enforces bound
// secrets on its own endpoints; this side relies on origin + paired tokens.

import type { Request, RequestHandler } from "express";
import crypto from "node:crypto";

const MUTATING = new Set(["POST", "PUT", "PATCH", "DELETE"]);

function timingSafeEqual(a: string, b: string): boolean {
  const ba = Buffer.from(a);
  const bb = Buffer.from(b);
  return ba.length === bb.length && crypto.timingSafeEqual(ba, bb);
}

function buildAllowlist(): Set<string> {
  const allowed = new Set<string>();
  const add = (raw?: string | null) => {
    if (!raw) return;
    try {
      const u = new URL(raw.trim());
      if ((u.protocol === "https:" || u.protocol === "http:") && u.host) {
        allowed.add(`${u.protocol}//${u.host}`);
      }
    } catch {
      /* ignore malformed entries */
    }
  };
  add(process.env.FRONTEND_URL);
  add(process.env.APP_URL);

  if (process.env.NODE_ENV !== "production") {
    for (const host of [
      "localhost:5173",
      "127.0.0.1:5173",
      "localhost:5000",
      "127.0.0.1:5000",
    ]) {
      allowed.add(`http://${host}`);
      allowed.add(`https://${host}`);
    }
  }
  return allowed;
}

const ALLOWED_ORIGINS = buildAllowlist();

export function isValidOrigin(
  req: Request,
  originOrReferer: string | undefined,
): boolean {
  // Exact scheme+host matching ONLY — substring checks are bypassable.
  if (!originOrReferer) return false;

  let url: URL;
  try {
    url = new URL(originOrReferer);
  } catch {
    return false;
  }

  if (req.get("host") && url.host === req.get("host")) return true; // same-origin

  let forwardedHost: string | undefined;
  try {
    forwardedHost =
      new URL(`${url.protocol}//${url.host}`).host ===
      req.get("x-forwarded-host")
        ? req.get("x-forwarded-host")
        : undefined;
  } catch {
    forwardedHost = undefined;
  }
  if (
    forwardedHost &&
    forwardedHost === req.get("host") &&
    url.host === forwardedHost
  ) {
    return true;
  }

  return ALLOWED_ORIGINS.has(`${url.protocol}//${url.host}`);
}

function cookieValue(req: Request, names: string[]): string | undefined {
  for (const name of names) {
    const v = req.cookies?.[name];
    if (typeof v === "string" && v.length > 0) return v;
  }
  return undefined;
}

function hasSessionCookie(req: Request): boolean {
  return Boolean(
    cookieValue(req, ["__Host-meshwork_session", "meshwork_session"]),
  );
}

export function originAllowed(req: Request): boolean {
  const origin = req.get("origin");
  const referer = req.get("referer");

  // Fail closed for cookie-bearing browser mutations with no origin signal.
  if (!origin && !referer && hasSessionCookie(req)) return false;
  if (req.get("Sec-Fetch-Site") && !origin && !referer) return false;
  if (origin) return isValidOrigin(req, origin);
  if (referer) return isValidOrigin(req, referer);
  return true; // plain API client without cookies cannot be CSRF'd
}

export const csrfProtect: RequestHandler = (req, res, next) => {
  if (!MUTATING.has(req.method)) return next();

  if (!originAllowed(req)) {
    return res
      .status(403)
      .json({ code: "CSRF_REJECTED", message: "CSRF validation failed" });
  }

  const cookieToken =
    cookieValue(req, ["__Host-meshwork_csrf", "meshwork_csrf"]) ?? "";
  const headerToken = req.get("X-CSRF-Token") ?? "";
  if (
    !cookieToken ||
    !headerToken ||
    !timingSafeEqual(cookieToken, headerToken)
  ) {
    return res
      .status(403)
      .json({ code: "CSRF_REJECTED", message: "CSRF validation failed" });
  }
  next();
};
